contrib/storage-splunk/src/main/java/org/apache/drill/exec/store/splunk/SplunkBatchWriter.java

```diff
@@ -98,27 +100,69 @@ public void updateSchema(VectorAccessible batch) {
   @Override
   public void startRecord() {
     logger.debug("Starting record");
-    // Ensure that the new record is empty. This is not strictly necessary, but it is a belt and suspenders approach.
-    splunkEvent.clear();
+    // Ensure that the new record is empty.
+    splunkEvent = new JSONObject();
   }
 
   @Override
-  public void endRecord() throws IOException {
+  public void endRecord() {
     logger.debug("Ending record");
+    recordCount++;
+
+    // Put event in buffer
+    eventBuffer.add(splunkEvent);
+
     // Write the event to the Splunk index
-    destinationIndex.submit(eventArgs, splunkEvent.toJSONString());
-    // Clear out the splunk event.
-    splunkEvent.clear();
+    if (recordCount >= config.getPluginConfig().getWriterBatchSize()) {
+      try {
+        writeEvents();
+      } catch (IOException e) {
+        throw  UserException.dataWriteError(e)
+            .message("Error writing data to Splunk: " + e.getMessage())
+            .build(logger);
+      }
+
+      // Reset record count
+      recordCount = 0;
+    }
   }
 
+
+  /*
+  args – Optional arguments for this stream. Valid parameters are: "host", "host_regex", "source", and "sourcetype".
+   */
   @Override
   public void abort() {
+    logger.debug("Aborting writing records to Splunk.");
     // No op
   }
 
   @Override
   public void cleanup() {
-    // No op
+    try {
+      writeEvents();
+    } catch (IOException e) {
+      throw  UserException.dataWriteError(e)
+          .message("Error writing data to Splunk: " + e.getMessage())
+          .build(logger);
+    }
+  }
+
+  private void writeEvents() throws IOException {
+    // Open the socket and stream, set up a timestamp
+    destinationIndex.attachWith(new ReceiverBehavior() {
```

Comment:
This results in a dedicated TCP socket being opened and closed for every writer batch.